### PR TITLE
Implement I2C

### LINKED ISSUE #51: Add I2C

### PR DESCRIPTION
**Pull request by Redrield**
Implement I2C based on the wpilibc I2C.cpp implementation. Fixes #51

### CODE REVIEW

**Comment by auscompgeek** [comment]
All the transaction calls are ioctl calls, and ioctl will return -1 and set errno on error. Best to expose the errno, I think.

I'm not sure what the return value is on success. It might be only 0 that is returned on success, or it could be the number of words transacted. Someone should probably eventually verify that.

Either way, definitely expose the errno. Take a look at the SPI PR if you're unsure.

File: wpilib/src/i2c.rs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+    /// Execute a write transaction with the device.
+    ///
+    /// Write a single byte to a register on a device and wait until the
+    ///   transaction is complete.
+    ///
+    /// Returns false if the transaction was successful, true if it was aborted
+    pub fn write(&self, register_address: i32, data: u8) -> bool {
+        let mut buf = [0u8; 2];
+        buf[0] = register_address as u8; // TODO: Is this valid?
```
Comment (auscompgeek) [inline wpilib/src/i2c.rs]: I2C registers can't be larger than a u8, so this is technically correct. Should take a u8 for the address though.

File: wpilib/src/i2c.rs [hunk 2]
```diff
@@ -63,10 +64,10 @@ impl I2C {
...
     ///   transaction is complete.
     ///
-    /// Returns false if the transaction was successful, true if it was aborted
-    pub fn write(&self, register_address: i32, data: u8) -> bool {
+    /// Returns a result based on whether the transaction was successful
+    pub fn write(&self, register_address: u8, data: u8) -> io::Result<usize> {
         let mut buf = [0u8; 2];
-        buf[0] = register_address as u8; // TODO: Is this valid?
+        buf[0] = register_address;
         buf[1] = data;
```
Comment (auscompgeek) [inline wpilib/src/i2c.rs]: Hm. Wouldn't it be better to construct the buf directly?

File: wpilib/src/i2c.rs [hunk 3]
```diff
@@ -107,10 +108,13 @@ impl I2C {
...
     /// transaction.
     ///
-    /// Returns false if the transfer was successful, true if it was aborted
-    pub fn read(&self, register_address: i32, buf: &mut [u8]) -> bool {
+    /// Returns a result based on whether the transaction was successful
+    pub fn read(&self, register_address: i32, buf: &mut [u8]) -> io::Result<usize> {
         if buf.is_empty() {
-            return true;
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
```
Comment (auscompgeek) [inline wpilib/src/i2c.rs]: Hm. Surely one of the existing ErrorKinds would be appropriate here.

File: wpilib/src/i2c.rs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+pub struct I2C {
+    port: Port,
+    device_address: i32,
+}
+
+impl I2C {
+    /// Constructs a new I2C
+    ///
+    /// `port` is the I2C port to which the device is connected, and `device_address` is the address of the device on the bus
+    pub fn new(port: Port, device_address: i32) -> HalResult<Self> {
```
Comment (auscompgeek) [inline wpilib/src/i2c.rs]: So I2C addresses are either 7 bits or 10 bits. I'm not sure whether the roboRIO does 10-bit addressing, but I guess if it (or a future platform) does the HAL would abstract it away. Make the input `device_address` a u16 I guess?

Comment (Redrield) [reply wpilib/src/i2c.rs]: All of the HAL functions expect `i32` for the device address field, and clippy doesn't like potentially lossy casts from u16 to i32

File: wpilib/src/i2c.rs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        loop {
+            if i >= expected.len() {
+                break;
+            }
+
+            let to_read = if expected.len() - i < 4 {
+                expected.len() - i
+            } else {
+                4
+            };
```
Comment (auscompgeek) [inline wpilib/src/i2c.rs]: What about this:
```rust
let to_read = 4.min(expected.len() - i);
```

File: wpilib/src/i2c.rs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            }
+
+            for j in 0..to_read {
+                if buf[j] != expected[i + j] {
+                    return false;
+                }
+            }
+
+            i += 4;
+            cur_register_address += 4;
```
Comment (auscompgeek) [inline wpilib/src/i2c.rs]: Surely there's a more idiomatic way to iterate by groups of 4 without having to directly index.

File: wpilib/src/i2c.rs [hunk 7]
```diff
@@ -151,11 +152,7 @@ impl I2C {
             .step_by(4)
             .zip((register_address..).step_by(4))
         {
-            let to_read = if expected.len() - i < 4 {
-                expected.len() - i
-            } else {
-                4
-            };
+            let to_read = cmp::min(4, expected.len() - i);
```
Comment (auscompgeek) [inline wpilib/src/i2c.rs]: Any particular reason here to not use Ord's min method directly?

Comment (Redrield) [reply wpilib/src/i2c.rs]: They're equivalent, I can switch it if it would be preferred

**Comment by Lytigas** [comment]
it's worth looking into the ioctl calls, but because WPILib only gives `bool` at the moment, we can be fairly confident its just success/failure. The HAL docs aren't specific on the meaning of a > 0 return.

**Pull request merged**